src/blog/fearless-structured-concurrency.vmd

```diff
@@ -0,0 +1,542 @@
+
+
+And just because we can, let's modify something created _inside_ the block:
+
+
+```
+fn main() export {
+  exponent = 3;
+
+  results =
+    parallel foreach 0..5 as i {
+      a = pow(i, exponent);
+      set a = a + 4; // We can modify!
+      = a; «blockresult2»
+    }
+}
+```
+
+You can see how the compiler can enforce that we only modify things that came from inside the block.
+
+
+The type system also needs to track what we can modify and what we can't. For that, we use `'r`.
+
+
+Note how in this snippet, `blur`'s first parameter's type is `'r Array<Array<int>>`. The `'r` enforces that the function won't modify the array from this function. [# This is similar to Rust's lifetimes, more on that below.]
+
+
+```
+fn main() export {
+  // Makes a 2D array of ints
+  input = Array(10, x => Array(10, y => x * 10 + y));
+
+  results =
+    parallel foreach 0..10 as x {
+      = map 0..10 as y {
+          = blur(input, x, y);
+        }
+    }
+}
+
+fn blur(input 'r &Array<Array<int>>, x int, y int) int {
+  ... // Loop over the 3x3 ints around x,y in input and return the average
+}
+```
+
+
+
+That `'r` means we're pointing into an *immutable region*. More on that below!
+
+
+<slice>
+#executor: The `parallel` block will by default spawn `num_cores * 2` threads, but it can also be specified via `parallel(num_threads)` or `parallel(an_executor)`.
+
+#blockresult1: The `=` here means this line produces the block's result.
+
+#blockresult2: The `=` here means this line produces the block's result.
+</slice>
+
+
+# Immutable References
+
+As stated above, to enable fearless structured concurrency, we need to express that a reference points into an *immutable region*.
+
+
+When an object is in an immutable region, we:
+
+ * Can't modify it.
+ * Can load a member (or element) from it.
+    * The type system remembers that it's also in the immutable region.
+ * Can pass it to a function, if the function knows it's from an immutable region.
+
+
+We're using `'r` to specify that a reference is in an immutable region. For example, an `'r &Array<int>` is an non-owning reference to an array of integers inside an immutable region.
+
+
+
+We can use these for our concurrency, with the following rule:
+
+> When inside a `parallel` statement, everything outside is in an immutable region.
+
+
+
+Elaborating on the `'r` a little more:
+
+ * The `r` is arbitrary, we can name it anything we want.
+ * The `'r` refers to a *scope*, or in other words, a *region*, similar to what Rust did above. [# This is similar to a lifetime in Rust, except that in Vale, mutability is per-region, not per-object: `'r` refers to an immutable region, `'r!` refers to a mutable region. There is no `&mut` in Vale. This adjustment makes the language easier to use, and enables fearless structured concurrency and other features.]
+ * By default, any named region is *immutable*.
+    * We can refer to a mutable region with `'r!` or just leave it unnamed.
```

Comment:
Put this in notes or something double nested bullets seems weird

Reply:
done